File: integrasjonspunkt/src/test/java/no/difi/meldingsutveksling/nextmove/v2/NextMoveValidatorTest.java
package no.difi.meldingsutveksling.nextmove.v2;

import no.difi.meldingsutveksling.config.AltinnFormidlingsTjenestenConfig;
import no.difi.meldingsutveksling.ServiceIdentifier;
import no.difi.meldingsutveksling.api.ConversationService;
import no.difi.meldingsutveksling.api.OptionalCryptoMessagePersister;
import no.difi.meldingsutveksling.arkivmelding.ArkivmeldingUtil;
import no.difi.meldingsutveksling.config.IntegrasjonspunktProperties;
import no.difi.meldingsutveksling.domain.sbdh.SBDService;
import no.difi.meldingsutveksling.domain.sbdh.SBDUtil;
import no.difi.meldingsutveksling.domain.sbdh.StandardBusinessDocument;
import no.difi.meldingsutveksling.exceptions.*;
import no.difi.meldingsutveksling.ks.svarut.SvarUtService;
import no.difi.meldingsutveksling.nextmove.*;
import no.difi.meldingsutveksling.serviceregistry.externalmodel.ServiceRecord;
import no.difi.meldingsutveksling.status.Conversation;
import no.difi.meldingsutveksling.validation.Asserter;
import no.difi.meldingsutveksling.validation.IntegrasjonspunktCertificateValidator;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.MockedStatic;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.beans.factory.ObjectProvider;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class NextMoveValidatorTest {

    @Mock
    private ServiceRecordProvider serviceRecordProvider;
    @Mock
    private NextMoveMessageOutRepository nextMoveMessageOutRepository;
    @Mock
    private ConversationStrategyFactory conversationStrategyFactory;
    @Mock
    private Asserter asserter;
    @Mock
    private OptionalCryptoMessagePersister optionalCryptoMessagePersister;
    @Mock
    private TimeToLiveHelper timeToLiveHelper;
    @Mock
    private SBDService sbdService;
    @Mock
    private ConversationService conversationService;
    @Mock
    private ArkivmeldingUtil arkivmeldingUtil;
    @Mock
    private NextMoveFileSizeValidator nextMoveFileSizeValidator;
    @Mock
    private IntegrasjonspunktProperties props;
    @Mock
    private ObjectProvider<IntegrasjonspunktCertificateValidator> certValidator;
    @Mock
    private SvarUtService svarUtService;
    @Mock
    private ObjectProvider<SvarUtService> svarUtServiceProvider;

    private NextMoveValidator nextMoveValidator;

    private final String messageId = "123";
    @Mock
    private NextMoveOutMessage message;
    @Mock
    private StandardBusinessDocument sbd;
    @Mock
    private ServiceRecord serviceRecord;

    private MockedStatic<SBDUtil> sbdUtilMock;

    private BusinessMessageFile bmf;

    private ArkivmeldingMessage businessMessage;

    @BeforeEach
    void before() {
        nextMoveValidator = new NextMoveValidator(
                serviceRecordProvider,
                nextMoveMessageOutRepository,
                conversationStrategyFactory,
                asserter,
                optionalCryptoMessagePersister,
                timeToLiveHelper,
                sbdService,
                conversationService,
                arkivmeldingUtil,
                nextMoveFileSizeValidator,
                props,
                certValidator,
                svarUtServiceProvider
        );

        bmf = new BusinessMessageFile()
                .setFilename("foo.txt")
                .setPrimaryDocument(true);

        businessMessage = new ArkivmeldingMessage().setHoveddokument("foo.txt");

        sbdUtilMock = mockStatic(SBDUtil.class);
    }

    @AfterEach
    void after() {
        if (sbdUtilMock != null) {
            sbdUtilMock.close();
        }
    }

    @Test
    void message_type_must_fit_document_type() {
        doAnswer(inv -> {
            ((java.util.function.Consumer<IntegrasjonspunktCertificateValidator>) inv.getArgument(0)).accept(mock(IntegrasjonspunktCertificateValidator.class));
            return null;
        }).when(certValidator).ifAvailable(any());
        when(sbd.getMessageId()).thenReturn(messageId);
        when(nextMoveMessageOutRepository.findByMessageId(messageId)).thenReturn(Optional.empty());
        when(conversationService.findConversation(messageId)).thenReturn(Optional.empty());
        when(serviceRecordProvider.getServiceIdentifier(sbd)).thenReturn(ServiceIdentifier.DPO);
        when(conversationStrategyFactory.isEnabled(ServiceIdentifier.DPO)).thenReturn(true);
        sbdUtilMock.when(() -> SBDUtil.isStatus(sbd)).thenReturn(false);
        sbdUtilMock.when(() -> SBDUtil.isReceipt(sbd)).thenReturn(false);
        sbdUtilMock.when(() -> SBDUtil.isFileRequired(sbd)).thenReturn(true);
        when(sbd.getDocumentType()).thenReturn("foo::bar");
        when(sbd.getType()).thenReturn("arkivmelding");

        assertThrows(MessageTypeDoesNotFitDocumentTypeException.class, () -> nextMoveValidator.validate(sbd));
    }

    @Test
    void document_type_must_be_valid() {
        when(sbd.getType()).thenReturn("melding");
        assertThrows(UnknownMessageTypeException.class, () -> nextMoveValidator.validate(sbd));
    }

    @Test
    void service_not_enabled_should_throw_exception() {
        doAnswer(inv -> {
            ((java.util.function.Consumer<IntegrasjonspunktCertificateValidator>) inv.getArgument(0)).accept(mock(IntegrasjonspunktCertificateValidator.class));
            return null;
        }).when(certValidator).ifAvailable(any());

        when(sbd.getMessageId()).thenReturn(messageId);
        when(sbd.getType()).thenReturn("arkivmelding");
        when(nextMoveMessageOutRepository.findByMessageId(messageId)).thenReturn(Optional.empty());
        when(conversationService.findConversation(messageId)).thenReturn(Optional.empty());
        when(serviceRecordProvider.getServiceIdentifier(sbd)).thenReturn(ServiceIdentifier.DPO);
        when(conversationStrategyFactory.isEnabled(ServiceIdentifier.DPO)).thenReturn(true);
        sbdUtilMock.when(() -> SBDUtil.isStatus(sbd)).thenReturn(false);
        sbdUtilMock.when(() -> SBDUtil.isReceipt(sbd)).thenReturn(false);
        sbdUtilMock.when(() -> SBDUtil.isFileRequired(sbd)).thenReturn(true);
        when(conversationStrategyFactory.isEnabled(ServiceIdentifier.DPO)).thenReturn(false);
        assertThrows(ServiceNotEnabledException.class, () -> nextMoveValidator.validate(sbd));
    }

    @Test
    void duplicate_messageId_not_allowed() {
        when(nextMoveMessageOutRepository.findByMessageId(messageId)).thenReturn(Optional.of(message));
        when(sbd.getMessageId()).thenReturn(messageId);
        sbdUtilMock.when(() -> SBDUtil.isStatus(sbd)).thenReturn(false);
        sbdUtilMock.when(() -> SBDUtil.isReceipt(sbd)).thenReturn(false);
        sbdUtilMock.when(() -> SBDUtil.isFileRequired(sbd)).thenReturn(true);
        assertThrows(MessageAlreadyExistsException.class, () -> nextMoveValidator.validate(sbd));
    }

    @Test
    void conversation_cannot_exist_with_same_messageId() {
        when(sbd.getMessageId()).thenReturn(messageId);
        when(nextMoveMessageOutRepository.findByMessageId(messageId)).thenReturn(Optional.empty());
        when(conversationService.findConversation(messageId)).thenReturn(Optional.empty());
        sbdUtilMock.when(() -> SBDUtil.isStatus(sbd)).thenReturn(false);
        sbdUtilMock.when(() -> SBDUtil.isReceipt(sbd)).thenReturn(false);
        sbdUtilMock.when(() -> SBDUtil.isFileRequired(sbd)).thenReturn(true);
        when(conversationService.findConversation(messageId)).thenReturn(Optional.of(mock(Conversation.class)));

        assertThrows(MessageAlreadyExistsException.class, () -> nextMoveValidator.validate(sbd));
    }

    @Test
    void non_receipt_messages_must_have_attachments() {
        doAnswer(inv -> {
            ((java.util.function.Consumer<IntegrasjonspunktCertificateValidator>) inv.getArgument(0)).accept(mock(IntegrasjonspunktCertificateValidator.class));
            return null;
        }).when(certValidator).ifAvailable(any());
        when(message.getSbd()).thenReturn(sbd);
        when(message.getFiles()).thenReturn(java.util.Collections.emptySet());
        sbdUtilMock.when(() -> SBDUtil.isStatus(sbd)).thenReturn(false);
        sbdUtilMock.when(() -> SBDUtil.isReceipt(sbd)).thenReturn(false);
        sbdUtilMock.when(() -> SBDUtil.isFileRequired(sbd)).thenReturn(true);
        when(message.getFiles()).thenReturn(null);

        assertThrows(MissingFileException.class, () -> nextMoveValidator.validate(message));
    }

    @Test
    void duplicate_filenames_not_allowed() {
        when(message.getOrCreateFiles()).thenReturn(new java.util.LinkedHashSet<>(java.util.Set.of(bmf)));
        sbdUtilMock.when(() -> SBDUtil.isStatus(sbd)).thenReturn(false);
        sbdUtilMock.when(() -> SBDUtil.isReceipt(sbd)).thenReturn(false);
        sbdUtilMock.when(() -> SBDUtil.isFileRequired(sbd)).thenReturn(true);

        BasicNextMoveFile file = BasicNextMoveFile.of("title", "foo.txt", "text", "foo".getBytes());
        assertThrows(DuplicateFilenameException.class, () -> nextMoveValidator.validateFile(message, file));
    }

    @Test
    void unknown_document_type_allowed_for_fiksio_message() {
        when(sbd.getDocumentType()).thenReturn("foo::bar");
        when(sbd.getType()).thenReturn("fiksio");
        when(serviceRecordProvider.getServiceIdentifier(any())).thenReturn(ServiceIdentifier.DPFIO);
        when(conversationStrategyFactory.isEnabled(ServiceIdentifier.DPFIO)).thenReturn(true);

        nextMoveValidator.validate(sbd);
    }

    @Test
    void dpo_message_does_not_require_title() {
        String filename = "bar.txt";
        when(message.isPrimaryDocument(filename)).thenReturn(false);
        BasicNextMoveFile file = BasicNextMoveFile.of("", filename, "text", "foo".getBytes());
        nextMoveValidator.validateFile(message, file);
    }

    @Test
    void dpv_message_requires_title() {
        String filename = "bar.txt";
        when(message.getServiceIdentifier()).thenReturn(ServiceIdentifier.DPV);
        when(message.isPrimaryDocument(filename)).thenReturn(false);
        BasicNextMoveFile file = BasicNextMoveFile.of("", "bar.txt", "text", "foo".getBytes());
        assertThrows(MissingFileTitleException.class, () -> nextMoveValidator.validateFile(message, file));
    }

    @Test
    void non_matching_channel_should_throw_exception() {
        AltinnFormidlingsTjenestenConfig dpo = mock(AltinnFormidlingsTjenestenConfig.class);
        when(props.getDpo()).thenReturn(dpo);
        when(dpo.getMessageChannel()).thenReturn("foo-42");

        when(sbd.getType()).thenReturn("arkivmelding");
        sbdUtilMock.when(() -> SBDUtil.getOptionalMessageChannel(sbd)).thenReturn(
                Optional.of(new no.difi.meldingsutveksling.domain.sbdh.Scope().setIdentifier("foo-43"))
        );

        doAnswer(inv -> {
            ((java.util.function.Consumer<IntegrasjonspunktCertificateValidator>) inv.getArgument(0)).accept(mock(IntegrasjonspunktCertificateValidator.class));
            return null;
        }).when(certValidator).ifAvailable(any());

        when(sbd.getMessageId()).thenReturn(messageId);
        when(sbd.getDocumentType()).thenReturn("standard::arkivmelding");
        when(nextMoveMessageOutRepository.findByMessageId(messageId)).thenReturn(Optional.empty());
        when(conversationService.findConversation(messageId)).thenReturn(Optional.empty());
        when(serviceRecordProvider.getServiceIdentifier(sbd)).thenReturn(ServiceIdentifier.DPO);
        when(conversationStrategyFactory.isEnabled(ServiceIdentifier.DPO)).thenReturn(true);
        sbdUtilMock.when(() -> SBDUtil.isStatus(sbd)).thenReturn(false);
        sbdUtilMock.when(() -> SBDUtil.isReceipt(sbd)).thenReturn(false);
        sbdUtilMock.when(() -> SBDUtil.isFileRequired(sbd)).thenReturn(true);

        assertThrows(MessageChannelInvalidException.class, () -> nextMoveValidator.validate(sbd));
    }
}

File: integrasjonspunkt/src/test/kotlin/no/difi/meldingsutveksling/nextmove/v2/NextMoveValidatorTest.kt
package no.difi.meldingsutveksling.nextmove.v2

import io.mockk.*
import io.mockk.impl.annotations.MockK
import lombok.extern.slf4j.Slf4j
import no.difi.meldingsutveksling.MessageType
import no.difi.meldingsutveksling.ServiceIdentifier
import no.difi.meldingsutveksling.api.ConversationService
import no.difi.meldingsutveksling.api.OptionalCryptoMessagePersister
import no.difi.meldingsutveksling.arkivmelding.ArkivmeldingUtil
import no.difi.meldingsutveksling.config.IntegrasjonspunktProperties
import no.difi.meldingsutveksling.domain.sbdh.SBDService
import no.difi.meldingsutveksling.domain.sbdh.SBDUtil
import no.difi.meldingsutveksling.domain.sbdh.StandardBusinessDocument
import no.difi.meldingsutveksling.exceptions.*
import no.difi.meldingsutveksling.ks.svarut.SvarUtService
import no.difi.meldingsutveksling.nextmove.*
import no.difi.meldingsutveksling.serviceregistry.externalmodel.ServiceRecord
import no.difi.meldingsutveksling.status.Conversation
import no.difi.meldingsutveksling.validation.Asserter
import no.difi.meldingsutveksling.validation.IntegrasjonspunktCertificateValidator
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.ObjectProvider
import java.util.*

@Slf4j
class NextMoveValidatorTest {

    @MockK
    lateinit var serviceRecordProvider: ServiceRecordProvider

    @MockK
    lateinit var nextMoveMessageOutRepository: NextMoveMessageOutRepository

    @MockK
    lateinit var conversationStrategyFactory: ConversationStrategyFactory

    @MockK
    lateinit var asserter: Asserter

    @MockK
    lateinit var optionalCryptoMessagePersister: OptionalCryptoMessagePersister

    @MockK
    lateinit var timeToLiveHelper: TimeToLiveHelper

    @MockK
    lateinit var sbdService: SBDService

    @MockK
    lateinit var conversationService: ConversationService

    @MockK
    lateinit var arkivmeldingUtil: ArkivmeldingUtil

    @MockK
    lateinit var nextMoveFileSizeValidator: NextMoveFileSizeValidator

    @MockK
    lateinit var props: IntegrasjonspunktProperties

    @MockK
    lateinit var certValidator: ObjectProvider<IntegrasjonspunktCertificateValidator>

    @MockK
    lateinit var svarUtService: SvarUtService

    @MockK
    lateinit var svarUtServiceProvider: ObjectProvider<SvarUtService>

    private lateinit var nextMoveValidator: NextMoveValidator

    private val messageId = "123"
    private val message = mockk<NextMoveOutMessage>()
    private val sbd = mockk<StandardBusinessDocument>()
    private val serviceRecord = mockk<ServiceRecord>()
    private val businessMessage = ArkivmeldingMessage()
        .setHoveddokument("foo.txt")

    @BeforeEach
    fun before() {
        MockKAnnotations.init(this)
        nextMoveValidator = NextMoveValidator(
            serviceRecordProvider,
            nextMoveMessageOutRepository,
            conversationStrategyFactory,
            asserter,
            optionalCryptoMessagePersister,
            timeToLiveHelper,
            sbdService,
            conversationService,
            arkivmeldingUtil,
            nextMoveFileSizeValidator,
            props,
            certValidator,
            svarUtServiceProvider
        )


        val bmf = BusinessMessageFile()
            .setFilename("foo.txt")
            .setPrimaryDocument(true)
        every { message.orCreateFiles } returns mutableSetOf(bmf)

        every { certValidator.ifAvailable(any()) } just Runs
        every { svarUtServiceProvider.getIfAvailable() } returns svarUtService
        every { svarUtServiceProvider.getObject() } returns svarUtService
        every { message.businessMessage } returns businessMessage

        every { sbd.messageId } returns messageId
        every { sbd.documentType } returns "standard::arkivmelding"
        every { sbd.process } returns "arkivmelding:administrasjon"
        every { sbd.type } returns "arkivmelding"

        every { message.messageId } returns messageId
        every { message.sbd } returns sbd
        every { message.serviceIdentifier } returns ServiceIdentifier.DPO
        every { message.files } returns emptySet()
        every { nextMoveMessageOutRepository.findByMessageId(messageId) } returns Optional.empty()
        every { conversationService.findConversation(messageId) } returns Optional.empty()
        every { serviceRecord.serviceIdentifier } returns ServiceIdentifier.DPO
        every { serviceRecordProvider.getServiceRecord(sbd) } returns serviceRecord
        every { serviceRecordProvider.getServiceIdentifier(sbd) } returns ServiceIdentifier.DPO
        every { conversationStrategyFactory.isEnabled(ServiceIdentifier.DPO) } returns true
        every { asserter.isValid(any<ArkivmeldingMessage>(), any()) } just Runs

        mockkStatic(SBDUtil::class)
        every { SBDUtil.isStatus(sbd) } returns false
        every { SBDUtil.isReceipt(sbd) } returns false
        every { SBDUtil.isFileRequired(sbd) } returns true
        every { nextMoveFileSizeValidator.validate(any(), any()) } just Runs
    }

    @AfterEach
    fun after() {
        clearStaticMockk(SBDUtil::class)
    }

    @Test
    fun `message type must fit document type`() {
        every { sbd.documentType } returns "foo::bar"
        every { sbd.type } returns "arkivmelding"
        assertThrows(MessageTypeDoesNotFitDocumentTypeException::class.java) { nextMoveValidator.validate(sbd) }
    }

    @Test
    fun `document type must be valid`() {
        every { sbd.type } returns "melding"
        assertThrows(UnknownMessageTypeException::class.java) { nextMoveValidator.validate(sbd) }
    }

    @Test
    fun `service not enabled should throw exception`() {
        every { conversationStrategyFactory.isEnabled(ServiceIdentifier.DPO) } returns false
        assertThrows(ServiceNotEnabledException::class.java) { nextMoveValidator.validate(sbd) }
    }

    @Test
    fun `duplicate messageId not allowed`() {
        every { nextMoveMessageOutRepository.findByMessageId(messageId) } returns Optional.of(message)
        assertThrows(MessageAlreadyExistsException::class.java) { nextMoveValidator.validate(sbd) }
    }

    @Test
    fun `conversation cannot exist with same messageId`() {
        every { conversationService.findConversation(messageId) } returns Optional.of(mockk<Conversation>())
        assertThrows(MessageAlreadyExistsException::class.java) { nextMoveValidator.validate(sbd) }
    }

    @Test
    fun `non-receipt messages must have attachments`() {
        every { message.files } returns null
        assertThrows(MissingFileException::class.java) { nextMoveValidator.validate(message) }
    }

    @Test
    fun `duplicate filenames not allowed`() {
        val file = BasicNextMoveFile.of("title", "foo.txt", "text", "foo".toByteArray())
        assertThrows(DuplicateFilenameException::class.java) { nextMoveValidator.validateFile(message, file) }
    }

    @Test
    fun `unknown document type allowed for fiksio message`() {
        every { sbd.documentType } returns "foo::bar"
        every { sbd.type } returns "fiksio"
        every { serviceRecord.serviceIdentifier } returns ServiceIdentifier.DPFIO
        every { serviceRecordProvider.getServiceIdentifier(any()) } returns ServiceIdentifier.DPFIO
        every { conversationStrategyFactory.isEnabled(ServiceIdentifier.DPFIO) } returns true
        every { sbd.any } returns mockkClass(FiksIoMessage::class)
        every { asserter.isValid(any<FiksIoMessage>(), any()) } just Runs

        nextMoveValidator.validate(sbd)
    }

    @Test
    fun `dpo message does not require title`() {
        val filename = "bar.txt"
        every { message.isPrimaryDocument(filename) } returns false
        val file = BasicNextMoveFile.of("", filename, "text", "foo".toByteArray())
        nextMoveValidator.validateFile(message, file)
    }

    @Test
    fun `dpv message requires title`() {
        val filename = "bar.txt"
        every { message.serviceIdentifier } returns ServiceIdentifier.DPV
        every { message.isPrimaryDocument(filename) } returns false
        val file = BasicNextMoveFile.of("", "bar.txt", "text", "foo".toByteArray())
        assertThrows(MissingFileTitleException::class.java) { nextMoveValidator.validateFile(message, file) }
    }

    @Test
    fun `non-matching channel should throw exception`() {
        every { props.dpo.messageChannel } returns "foo-42"
        every { sbd.type } returns "arkivmelding"
        every { SBDUtil.getOptionalMessageChannel(sbd) } returns Optional.of(mockk {
            every { identifier } returns "foo-43"
        })
        assertThrows(MessageChannelInvalidException::class.java) { nextMoveValidator.validate(sbd) }
    }

    @Test
    fun `print address without adresselinje 1 throws exception`() {
        val dpiPrintMessage = mockk<DpiPrintMessage>()
        val mottaker = mockk<PostAddress>()
        every { sbd.type } returns "print"
        every { sbd.any } returns dpiPrintMessage
        every { mottaker.adresselinje1 } returns null
        every { mottaker.postnummer } returns "5134"
        every { mottaker.poststed } returns "Flaktveit"
        every { mottaker.land } returns "Bergen"
        every { mottaker.landkode } returns "NO"
        every { mottaker.land } returns "norge"
        every {dpiPrintMessage.mottaker} returns mottaker

        assertThrows(MissingAddressInformationException::class.java) { nextMoveValidator.validate(sbd) }
    }

    @Test
    fun `print address with adresselinje1 but not (norwegian address || foreign address country || foreign address country code) exception`() {
        val dpiPrintMessage = mockk<DpiPrintMessage>()
        val mottaker = mockk<PostAddress>()
        every { sbd.type } returns "print"
        every { sbd.any } returns dpiPrintMessage
        every { mottaker.adresselinje1 } returns "adresselinje 1 korrekt utfylt"
        every { mottaker.postnummer } returns null
        every { mottaker.poststed } returns null
        every { mottaker.land } returns null
        every { mottaker.landkode } returns null
        every { dpiPrintMessage.mottaker} returns mottaker

        assertThrows(MissingAddressInformationException::class.java) { nextMoveValidator.validate(sbd) }
    }

    @Test
    fun `print address with adresselinje1 and Norway but no postal number exception`() {
        val dpiPrintMessage = mockk<DpiPrintMessage>()
        val mottaker = mockk<PostAddress>()
        every { sbd.type } returns "print"
        every { sbd.any } returns dpiPrintMessage
        every { mottaker.adresselinje1 } returns "adresselinje 1 korrekt utfylt"
        every { mottaker.postnummer } returns null
        every { mottaker.poststed } returns null
        every { mottaker.land } returns "norway"
        every { mottaker.landkode } returns null
        every { dpiPrintMessage.mottaker} returns mottaker

        assertThrows(MissingAddressInformationException::class.java) { nextMoveValidator.validate(sbd) }
    }

}